### PR TITLE
Adding GEP 3171: Percentage-based Request Mirroring

### PR DESCRIPTION
**Pull request by jakebennert**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
This adds a new GEP proposing a new feature Percentage-based Request Mirroring.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: geps/gep-3171/index.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        // MirrorPercent represents the fraction of requests that should be
+        // mirrored to BackendRef.
+        //
+        // If MirrorPercent is not specified, then 100% of requests will be
+        // mirrored.
+        //
+        // +optional
+        MirrorPercent Percentage `json:"mirrorPercent,omitempty"`
+}
+```
```
Comment (arkodg) [inline geps/gep-3171/index.md:85]: can you add an example ?

File: geps/gep-3171/index.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+## Goals
+
+Successfully implement the feature.
+
+## Introduction
+
+[Request Mirroring](https://gateway-api.sigs.k8s.io/guides/http-request-mirroring/) is a feature that allows a user to mirror requests going to some backend A along to some other specified backend B. Right now Request Mirroring is an all or nothing feature – either 100% of request are mirrored, or 0% are. Percentage-based Request Mirroring will allow users to specify a percentage of requests they'd like mirrored as opposed to every single request.   
+
+This feature is already [supported by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-routeaction-requestmirrorpolicy), so adding it for the Gateway API would enable better integration between the two products. There's also an existing user desire for this feature on the [HAProxy side](https://www.haproxy.com/blog/haproxy-traffic-mirroring-for-real-world-testing) and [NGINX side](https://alex.dzyoba.com/blog/nginx-mirror/). Since Request Mirroring is already supported by the Gateway API, Percentage-based Request Mirroring would a clear improvement on this pre-existing feature.
+
```
Comment (arkodg) [inline geps/gep-3171/index.md:21]: can we add a table highlighting past work ? for e.g. https://gateway-api.sigs.k8s.io/geps/gep-91/#existing-support-in-implementations

**Comment by candita** [comment]
I'd like to point out that we already have a concept of Weight that works on percentages, and it is a shared type. 
 Can it be reused here?  
https://github.com/kubernetes-sigs/gateway-api/blob/c8fd84d2369ba7d07155dc2e280586dc5601ffd9/apis/v1/shared_types.go#L268-L286

**Comment by kflynn** [comment]
@candita, I think BackendRef's weight definition won't well for mirroring, sadly. BackendRef's weight scheme is almost the inverse of what we want: it relies on you to put large weights on the refs that should get a _lot_ of traffic, and the only way to describe a small amount of traffic is to put large weights on all the _other_ refs.

This is problematic with mirroring, where we'd really like a really easy way to set up mirroring of a very small fraction of traffic _without_ having to edit all the resources describing the critical production paths. Overall, I think we really are going to need an arbitrary-fraction type.

**Comment by robscott** [comment]
THanks @jakebennert!

/ok-to-test

**Comment by candita** [comment]
@kflynn 
> it relies on you to put large weights on the refs that should get a _lot_ of traffic, and the only way to describe a small amount of traffic is to put large weights on all the _other_ refs.

Did you mean "the only way to describe a small amount of traffic is to put **small** weights on all the other refs"?  It looks like the default `Weight` value is 1, so as long as you put something higher than 1 on the refs that should be used more often, that would work.

e.g. if we have four backendRefs, and one of them we want to have a majority of all traffic.  From the godoc, it follows that you can leave 3 `Weight`s unspecified (defaults to 1), and on the fourth,  set `Weight` to 5.  The fourth one will get 5/8 of the traffic, and the other 3 will get 1/8 of the traffic.  Am I misinterpreting the godoc?

**Comment by kflynn** [comment]
@candita, in my experience, what I would often see happen is that the app devs will have a running workload, _then_ decide that they want to enable shadowing for some reason, which means that they're in the situation where they want to add something that gets a small fraction of traffic _after_ they've already set up the normal backends. It's much more user-friendly to add a single resource for "here's a backend that I want to shadow a tiny fraction of traffic" than to have to edit all your existing backends to add a large weight to them, then add a new resource that asks for a small fraction of traffic.

Also, suppose you currently have three backends with weights of 50%, 25%, and 25%, you get 50K RPS, and you want to shadow no more than 50RPS. Quick, what weights do you need to use to specify that? Should you re-edit the weights back to 50/25/25 when you turn off shadowing a week later? How about if you decide the shadow needs to drop to 25RPS? Do we start telling people that the best practice is to deploy with a weight of 10000 for all their backends in case they need to mess with stuff like this? 😐 

Let's just sidestep all those questions by having a way to say "here's a backend, give it 1/1000th of the traffic to this route, please." 🙂

**Comment by jakebennert** [comment]
> Love the idea! I personally would vote for allowing two forms:
> 
> ```
> MirrorFraction Fraction `json:"mirrorFraction,omitempty"`
> MirrorPercent int `json:"mirrorPercent,omitempty"`
> ```
> 
> with validation allowing only one to be set (and we can define that if they are somehow both set, `mirrorFraction` takes precedence because it's more precise).
> 
> This is slightly harder on implementers, but I expect that the vast majority of users will only need an integral percentage, and forcing all of them to write
> 
> ```yaml
> mirrorFraction:
>   numerator: 5
> ```
> 
> instead of
> 
> ```yaml
> mirrorPercent: 5
> ```
> 
> feels like putting the difficulty on the wrong users. It's definitely important to support the case where you have enough traffic that you need to mirror 1/10000th of it, but I don't think that's the common case, and I'd rather the common case be simple. 🙂

@kflynn I think this is a good idea. I've updated my proposal - how does this look to you?

**Comment by kflynn** [comment]
@jakebennert Looks great to me, thank you!! 🙂 

/lgtm

File: geps/gep-3171/index.md (hunk 1 above)
Comment (jakebennert) [reply geps/gep-3171/index.md:103]: How does this look to you?

File: geps/gep-3171/index.md (hunk 2 above)
Comment (jakebennert) [reply geps/gep-3171/index.md:21]: Done!

File: geps/gep-3171/index.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
+# GEP-3171: Percentage-based Request Mirroring
+
+* Issue: [#3171](https://github.com/kubernetes-sigs/gateway-api/issues/3171)
+* Status: **Provisional**
```
Comment (robscott) [inline geps/gep-3171/index.md:4]:
```suggestion
* Status: **Implementable**
```

**Review by robscott — COMMENTED** [review]
Thanks @jakebennert! One tiny nit about status, but otherwise LGTM. Would appreciate another LGTM from someone else.

/cc @arkodg @candita @kflynn @mlavacca

File: geps/gep-3171/index.md (hunk 3 above)
Comment (kflynn) [reply geps/gep-3171/index.md:4]: Nice catch, @robscott -- once this change is made I'll happily add an LGTM. 🙂

**Comment by jakebennert** [comment]
> Thanks @jakebennert! One tiny nit about status, but otherwise LGTM. Would appreciate another LGTM from someone else.
> 
> /cc @arkodg @candita @kflynn @mlavacca

@kflynn I know you LGTM'd in a previous comment, but I think because I made some commits after the fact, it's out of date. If it all looks good to you, would you be able to LGTM again? :)

**Comment by robscott** [comment]
Thanks @jakebennert! Will defer to @kflynn (or anyone else) for LGTM

/approve
/hold cancel

**Comment by kflynn** [comment]
/lgtm

again 🙂

**Pull request merged**